src/webots/nodes/WbCadShape.cpp

```diff
@@ -165,10 +226,83 @@ void WbCadShape::updateUrl() {
       }
     }
 
+    const QString extension = completeUrl.mid(completeUrl.lastIndexOf('.') + 1).toLower();
+    if (extension == "obj" && WbUrl::isWeb(completeUrl)) {
+      // ensure any mtl referenced by the obj file are also downloaded
+      if (areMaterialAssetsAvailable(completeUrl)) {
+        mObjMaterials.clear();
+        // generate mapping between referenced files and cached files
+        QStringList rawMaterials = objMaterialList(completeUrl);
+        foreach (QString material, rawMaterials) {
+          QString adjustedUrl = generateMaterialUrl(material, completeUrl);
+          assert(WbNetwork::instance()->isCached(adjustedUrl));
+          if (!mObjMaterials.contains(material))
+            mObjMaterials.insert(material, adjustedUrl);
+        }
+      } else {
+        retrieveMaterials();
+        return;
+      }
+    }
+
     createWrenObjects();
   }
 }
 
+bool WbCadShape::areMaterialAssetsAvailable(const QString &url) {
+  QStringList rawMaterials = objMaterialList(url);  // note: 'dae' files will generate an empty list
+  foreach (QString material, rawMaterials) {
+    if (!WbNetwork::instance()->isCached(generateMaterialUrl(material, url)))
+      return false;
+  }
+  return true;
+}
+
+void WbCadShape::generateMaterialMap(const QString &url) {
```

Comment:
It seems like this function is not used?

Reply:
Ops, indeed forgot to remove it after refactoring